src/rust/ide/src/controller/graph.rs

```diff
@@ -0,0 +1,218 @@
+pub struct Node {
+    /// Information based on AST, from double_representation module.
+    pub info : double_representation::node::NodeInfo,
+    /// Information about this node stored in the module's metadata.
+    pub metadata : NodeMetadata
+}
+
+
+
+// ===================
+// === NewNodeInfo ===
+// ===================
+
+/// Describes the node to be added.
+#[derive(Clone,Debug)]
+pub struct NewNodeInfo {
+    /// Expression to be placed on the node
+    pub expression : String,
+    /// Visual node position in the graph scene.
+    pub metadata : NodeMetadata,
+    /// ID to be given to the node.
+    pub id : Option<ast::ID>,
+    /// Where line created by adding this node should appear.
+    pub location_hint : LocationHint
+}
+
+/// Describes the desired position of the node's line in the graph's code block.
+#[derive(Clone,Copy,Debug)]
+pub enum LocationHint {
+    /// Try placing this node's line before the line described by id.
+    Before(ast::ID),
+    /// Try placing this node's line after the line described by id.
+    After(ast::ID),
+    /// Try placing this node's line at the start of the graph's code block.
+    Start,
+    /// Try placing this node's line at the end of the graph's code block.
+    End,
+}
+
+
+
+// ==================
+// === Controller ===
+// ==================
+
+/// State data of the module controller.
+#[derive(Clone,Debug)]
+pub struct Handle {
+    /// Controller of the module which this graph belongs to.
+    module    : controller::module::Handle,
+    id        : Id,
+    /// Publisher. When creating a controller, it sets up task to emit notifications through this
+    /// publisher to relay changes from the module controller.
+    publisher : Rc<RefCell<controller::notification::Publisher<controller::notification::Graph>>>,
+}
+
+impl Handle {
+    /// Gets a handle to a controller of the module that this definition belongs to.
+    pub fn get_module(&self) -> controller::module::Handle {
+        self.module.clone()
+    }
+
+    /// Gets a handle to a controller of the module that this definition belongs to.
+    pub fn get_id(&self) -> Id {
+        self.id.clone()
+    }
+
+    /// Creates a new controller. Does not check if id is valid.
+    ///
+    /// Requires global executor to spawn the events relay task.
+    pub fn new_unchecked(module:controller::module::Handle, id:Id) -> Handle {
+        let graphs_notifications = module.subscribe_graph_notifications();
+        let publisher = default();
+        let ret       = Handle {module,id,publisher};
+        let weak      = Rc::downgrade(&ret.publisher);
+        let relay_notifications = process_stream_with_handle(
+            graphs_notifications,
+            weak,
+            |notification,this| {
+                match notification {
+                    controller::notification::Graphs::Invalidate =>
+                    this.borrow_mut().publish(controller::notification::Graph::Invalidate),
+            }
+        });
```

Comment:
```suggestion
        let relay_notifications = process_stream_with_handle(graphs_notifications,weak,
            |notification,this| {
                match notification {
                    controller::notification::Graphs::Invalidate =>
                    this.borrow_mut().publish(controller::notification::Graph::Invalidate),
            }
        });
```
We rarely put arguments in a separate lines.

Additionally, I would rename "this" to "publisher" inside lambda